crates/chain/src/canonical_iter.rs

```diff
@@ -189,13 +203,30 @@ impl<'g, A: Anchor, C: ChainOracle> CanonicalIter<'g, A, C> {
 
         if detected_self_double_spend {
             for txid in staged_queue {
-                self.canonical.remove(&txid);
+                if let Some((tx, _)) = self.canonical.remove(&txid) {
+                    // Remove all the spends that were added for this transaction
+                    for input in &tx.input {
+                        self.canonical_spends.remove(&input.previous_output);
+                    }
+                }
             }
             for txid in undo_not_canonical {
                 self.not_canonical.remove(&txid);
             }
         } else {
-            self.queue.extend(staged_queue);
+            for txid in staged_queue {
```

Comment:
I don't think there is a use for `staged_queue`. Have `TxAncestors` return `Option<()>`. Insert into `canonical_roots` in the `TxAncestors` closure - we are already iterating through the inputs there anyway (to populate `canonical_spends`).